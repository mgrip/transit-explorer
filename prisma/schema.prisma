// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Route {
  id            Int            @id @default(autoincrement())
  name          String
  longName      String
  description   String?
  type          String?
  color         String?
  routePatterns RoutePattern[]
  sourceInfo    Json
  sourceId      String
}

// for now, we are only dealing with "canonical" route patterns
model RoutePattern {
  id                Int                @id @default(autoincrement())
  name              String
  route             Route              @relation(fields: [routeId], references: [id])
  routeId           Int
  routePatternStops RoutePatternStop[]
  sourceInfo        Json
  sourceId          String
}

model Stop {
  id                Int                @id @default(autoincrement())
  name              String
  description       String?
  address           String?
  latitude          Float
  longitude         Float
  routePatternStops RoutePatternStop[]
  sourceInfo        Json
  sourceId          String
}

model RoutePatternStop {
  routePattern   RoutePattern @relation(fields: [routePatternId], references: [id])
  routePatternId Int
  stop           Stop         @relation(fields: [stopId], references: [id])
  stopId         Int
  index          Int

  @@unique([routePatternId, stopId])
}
