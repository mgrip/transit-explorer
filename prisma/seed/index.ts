import { readFile } from "fs/promises";
import prisma from "../client.ts";

async function main() {
  const jsonString = await readFile(
    "prisma/seed/data/transit-data.json",
    "utf-8"
  );
  const seedData = JSON.parse(jsonString);

  await prisma.route.createMany({ data: seedData.routes });
  await prisma.stop.createMany({ data: seedData.stops });
  await prisma.routePattern.createMany({ data: seedData.routePatterns });
  await prisma.routePatternStop.createMany({
    data: seedData.routePatternStops,
  });
}

main()
  .then(async () => {
    await prisma.$disconnect();
  })
  .catch(async (e) => {
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });
