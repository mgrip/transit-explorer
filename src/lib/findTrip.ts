import prisma from "../../prisma/client.ts";

interface TripSegment {
  routePatternId: number;
  routeId: number;
  stopIds: number[];
}

type Trip = TripSegment[];

interface RoutePatternMap {
  id: number;
  routeId: number;
  routePatternStops: { id: number; stopId: number; index: number }[];
}
interface RouteMap {
  [routePatternId: number]: RoutePatternMap;
}

function incrementTripSegments(
  currentTrips: Trip[],
  routeMap: RouteMap
): Trip[] {
  return currentTrips.reduce<Trip[]>((acc, trip) => {
    const mostRecentSegment = trip.at(-1);
    if (!mostRecentSegment) {
      throw new Error("No current segment for trip");
    }
    const currentRoutePatternStop = routeMap[
      mostRecentSegment.routePatternId
    ].routePatternStops.find(
      (routePatternStop) =>
        routePatternStop.stopId === mostRecentSegment.stopIds.at(-1)
    );
    if (!currentRoutePatternStop) {
      console.log(mostRecentSegment, "1");
      console.log(routeMap[mostRecentSegment.routePatternId], "2");
      console.log(currentTrips[1][0].stopIds);
      throw new Error("Current route pattern does not include current stop");
    }

    const nextRoutePatternStop = routeMap[
      mostRecentSegment.routePatternId
    ].routePatternStops.find(
      (routePatternStop) =>
        routePatternStop.index === currentRoutePatternStop.index + 1
    );

    const availableTransferRoutePatterns = Object.values<RoutePatternMap>(
      // eslint-disable-next-line
      // @ts-ignore
      routeMap
    ).reduce<RoutePatternMap[]>((acc, routePattern) => {
      const matchingRoutePatternStop = routePattern.routePatternStops.find(
        (routePatternStop) =>
          routePatternStop.stopId === currentRoutePatternStop.stopId
      );

      // don't allow transfer to route patterns on the same route
      if (routePattern.routeId === mostRecentSegment.routeId) {
        return [...acc];
      } else if (matchingRoutePatternStop) {
        // check to match sure there is a next stop on any available transfer route pattern
        const nextStopOnTransferRoute = routePattern.routePatternStops.find(
          (routePatternStop) =>
            routePatternStop.index === matchingRoutePatternStop.index + 1
        );
        if (nextStopOnTransferRoute) {
          return [...acc, routePattern];
        } else {
          return [...acc];
        }
      } else {
        return [...acc];
      }
    }, []);

    let continuingTrips: Trip[] = [];
    if (nextRoutePatternStop && mostRecentSegment) {
      continuingTrips = [
        [
          ...trip.slice(0, -1),
          {
            ...mostRecentSegment,
            stopIds: [
              ...mostRecentSegment.stopIds,
              nextRoutePatternStop.stopId,
            ],
          },
        ],
      ];
    }

    let newTransferTrips: Trip[] = [];
    if (availableTransferRoutePatterns.length > 0) {
      newTransferTrips = availableTransferRoutePatterns.map(
        (transferRoutePattern) => {
          const matchingRoutePatternStop =
            transferRoutePattern.routePatternStops.find(
              (routePatternStop) =>
                routePatternStop.stopId === currentRoutePatternStop.stopId
            );
          if (!matchingRoutePatternStop) {
            throw new Error("should never get here");
          }
          const nextStopOnTransferRoute =
            transferRoutePattern.routePatternStops.find(
              (routePatternStop) =>
                routePatternStop.index === matchingRoutePatternStop.index + 1
            );
          if (!nextStopOnTransferRoute) {
            throw new Error("should never get here");
          }
          const newTrip: Trip = [
            ...trip,
            {
              routePatternId: transferRoutePattern.id,
              routeId: transferRoutePattern.routeId,
              stopIds: [
                currentRoutePatternStop.stopId,
                nextStopOnTransferRoute.stopId,
              ],
            },
          ];
          return newTrip;
        }
      );
    }

    // at some point its not worth checking the exponential number of possible trips based on every possible transfer combination
    // we're using a heuristic here to assume any trip with more than 3 segments is not worth checking
    const filterUnrealisticTrips = (trip: Trip) => trip.length < 4;

    return [
      ...acc,
      ...continuingTrips.filter(filterUnrealisticTrips),
      ...newTransferTrips.filter(filterUnrealisticTrips),
    ];
  }, []);
}

function checkForRoutes(
  currentTrips: Trip[],
  routeMap: RouteMap,
  destinationStopId: number,
  depth = 1000
): Trip {
  const newTrips = incrementTripSegments(currentTrips, routeMap);
  const completeTrips = newTrips.filter(
    (trip) => trip.at(-1)?.stopIds.at(-1) === destinationStopId
  );
  if (completeTrips.length > 0) {
    // sort by stop length
    completeTrips.sort((a, b) => {
      const aLength = a.reduce(
        (acc, segment) => acc + segment.stopIds.length,
        0
      );
      const bLength = b.reduce(
        (acc, segment) => acc + segment.stopIds.length,
        0
      );
      return aLength < bLength ? -1 : aLength > bLength ? 1 : 0;
    });

    return completeTrips[0];
  } else if (depth === 0) {
    throw new Error("Max depth exceeded - no route found");
  } else {
    return checkForRoutes(newTrips, routeMap, destinationStopId, depth - 1);
  }
}

export default async function findTrip({
  departureStopId,
  arrivalStopId,
}: {
  departureStopId: number;
  arrivalStopId: number;
}): Promise<Trip | undefined> {
  const routeMap: RouteMap = {};
  const routePatterns = await prisma.routePattern.findMany({
    include: {
      routePatternStops: true,
    },
  });
  routePatterns.forEach((routePattern) => {
    routeMap[routePattern.id] = {
      id: routePattern.id,
      routeId: routePattern.routeId,
      routePatternStops: routePattern.routePatternStops.map(
        (routePatternStop) => ({
          id: routePatternStop.index,
          stopId: routePatternStop.stopId,
          index: routePatternStop.index,
        })
      ),
    };
  });
  const availableRoutePatterns = await prisma.routePattern.findMany({
    where: {
      routePatternStops: {
        some: {
          stopId: departureStopId,
        },
      },
    },
  });
  const startTrips = availableRoutePatterns.map((routePattern) => {
    const trip: Trip = [
      {
        routeId: routePattern.routeId,
        routePatternId: routePattern.id,
        stopIds: [departureStopId],
      },
    ];
    return trip;
  });

  return checkForRoutes(startTrips, routeMap, arrivalStopId);
}
