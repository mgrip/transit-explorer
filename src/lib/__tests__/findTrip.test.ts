import { objectContainsKey } from "jest-mock-extended";
import { prismaMock } from "../../../prisma/singleton";
import { RoutePattern, RoutePatternStop } from "@prisma/client";
import findTrip from "../findTrip";

test("can find trip", async () => {
  type RoutePatternResult = RoutePattern & {
    routePatternStops: RoutePatternStop[];
  };
  const routePattern1: RoutePatternResult = {
    id: 1,
    name: "Route 1",
    routeId: 1,
    sourceInfo: {},
    sourceId: "1",
    routePatternStops: [
      { routePatternId: 1, stopId: 1, index: 0 },
      { routePatternId: 1, stopId: 2, index: 1 },
      { routePatternId: 1, stopId: 3, index: 2 },
    ],
  };
  const routePattern2: RoutePatternResult = {
    id: 2,
    name: "Route 2",
    routeId: 2,
    sourceInfo: {},
    sourceId: "2",
    routePatternStops: [
      { routePatternId: 2, stopId: 4, index: 0 },
      { routePatternId: 2, stopId: 3, index: 1 },
      { routePatternId: 2, stopId: 5, index: 2 },
    ],
  };
  const routePatterns = [routePattern1, routePattern2];

  prismaMock.routePattern.findMany
    // this is the call to get all route patterns
    .calledWith(objectContainsKey("include"))
    .mockResolvedValue(routePatterns);

  prismaMock.routePattern.findMany
    // this is the call to get routePatterns "where stop id = departure stop id",
    // so we want to just return the first route pattern which actually includes stop 1
    .calledWith(objectContainsKey("where"))
    .mockResolvedValue([routePattern1]);

  const result = await await findTrip({ departureStopId: 1, arrivalStopId: 5 });

  expect(result).toMatchSnapshot();
});
