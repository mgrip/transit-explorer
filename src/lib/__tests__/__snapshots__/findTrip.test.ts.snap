// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can find trip 1`] = `
[
  {
    "routeId": 1,
    "routePatternId": 1,
    "stopIds": [
      1,
      2,
      3,
    ],
  },
  {
    "routeId": 2,
    "routePatternId": 2,
    "stopIds": [
      3,
      5,
    ],
  },
]
`;
